File: cloud-run/translate/llm.js
const fs = require("fs");
const path = require("path");

const Anthropic = require("@anthropic-ai/sdk");
const { AnthropicVertex } = require("@anthropic-ai/vertex-sdk");
const { GoogleGenAI } = require("@google/genai");
const { StatusCodes } = require("http-status-codes");

require("dotenv").config({ path: path.resolve(__dirname, "../../.env") });

const PROVIDER_ANTHROPIC = "anthropic";
const PROVIDER_GOOGLE = "google";

// Backends: Vertex AI (service account auth, no API keys) or the direct
// provider APIs (API keys from Secret Manager). LLM_BACKEND picks the order;
// vertex-first falls back to direct while the Vertex migration is verified.
const BACKEND_VERTEX = "vertex";
const BACKEND_DIRECT = "direct";
const BACKEND_ORDER = {
  "vertex-first": [BACKEND_VERTEX, BACKEND_DIRECT],
  "vertex-only": [BACKEND_VERTEX],
  "direct-only": [BACKEND_DIRECT],
};
const DEFAULT_BACKEND_MODE = "vertex-first";
const DEFAULT_VERTEX_LOCATION = "us";

// Fall back only on errors raised before any generation work (auth, model not
// enabled/available, quota). Timeouts, 5xx, and bad output are not retried on
// the other backend: they can come after minutes of work and would double
// cost and latency while hiding the real failure.
const FALLBACK_STATUS_CODES = new Set([
  StatusCodes.UNAUTHORIZED,
  StatusCodes.FORBIDDEN,
  StatusCodes.NOT_FOUND,
  StatusCodes.TOO_MANY_REQUESTS,
]);
// google-auth-library throws plain Errors when no credentials can be found
const MISSING_CREDENTIALS_PATTERN = /default credentials/i;

class VertexNotConfiguredError extends Error {}

const SCHEMA_DIR = __dirname;
const SCHEMA_PATHS = {
  [PROVIDER_ANTHROPIC]: path.join(SCHEMA_DIR, "translation-schema-claude.json"),
  [PROVIDER_GOOGLE]: path.join(SCHEMA_DIR, "translation-schema-gemini.json"),
};

const DEFAULT_MAX_TOKENS = 65536;
const LLM_TIMEOUT_MS = 600_000;

function loadTranslationSchema(provider) {
  const schemaPath = SCHEMA_PATHS[provider];
  if (!schemaPath) {
    throw new Error(`No translation schema for provider: ${provider}`);
  }
  return JSON.parse(fs.readFileSync(schemaPath, "utf-8"));
}

function vertexSettings() {
  const project = process.env.VERTEX_PROJECT_ID;
  if (!project) {
    throw new VertexNotConfiguredError("VERTEX_PROJECT_ID is not set");
  }
  return { project, location: process.env.VERTEX_LOCATION || DEFAULT_VERTEX_LOCATION };
}

function makeClaudeClient(backend) {
  if (backend === BACKEND_VERTEX) {
    const { project, location } = vertexSettings();
    return new AnthropicVertex({ projectId: project, region: location, timeout: LLM_TIMEOUT_MS });
  }
  return new Anthropic({ timeout: LLM_TIMEOUT_MS });
}

function makeGeminiClient(backend) {
  if (backend === BACKEND_VERTEX) {
    const { project, location } = vertexSettings();
    return new GoogleGenAI({ vertexai: true, project, location });
  }
  return new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
}

/**
 * Why this error justifies retrying on the direct API, or null if it doesn't.
 * Both SDKs' API errors carry a numeric HTTP `status`.
 */
function fallbackReason(err) {
  if (err instanceof VertexNotConfiguredError) {
    return "vertex not configured";
  }
  if (typeof err?.status === "number" && FALLBACK_STATUS_CODES.has(err.status)) {
    return `HTTP ${err.status}: ${err.name || err.constructor.name}`;
  }
  if (MISSING_CREDENTIALS_PATTERN.test(err?.message || "")) {
    return "credentials: no default credentials";
  }
  return null;
}

function backendOrder() {
  const mode = process.env.LLM_BACKEND || DEFAULT_BACKEND_MODE;
  if (!BACKEND_ORDER[mode]) {
    throw new Error(`Unknown LLM_BACKEND: ${mode}`);
  }
  return BACKEND_ORDER[mode];
}

async function callClaude(client, prompt, { model, maxTokens = DEFAULT_MAX_TOKENS, outputSchema, effort } = {}) {
  const kwargs = {
    model,
    max_tokens: maxTokens,
    messages: [{ role: "user", content: prompt }],
  };

  if (outputSchema || effort) {
    kwargs.output_config = {};
    if (outputSchema) {
      kwargs.output_config.format = {
        type: "json_schema",
        schema: outputSchema,
      };
    }
    if (effort) {
      kwargs.output_config.effort = effort;
    }
  }

  const response = await client.messages.create(kwargs);
  const usage = {
    input_tokens: response.usage.input_tokens,
    output_tokens: response.usage.output_tokens,
  };
  const textBlock = response.content.find((b) => b.type === "text");
  return { text: textBlock.text, usage, stop_reason: response.stop_reason };
}

async function callGemini(client, prompt, { model, outputSchema } = {}) {
  const kwargs = {
    model,
    contents: [{ role: "user", parts: [{ text: prompt }] }],
  };

  if (outputSchema) {
    kwargs.config = {
      responseMimeType: "application/json",
      responseSchema: outputSchema,
    };
  }

  const response = await client.models.generateContent(kwargs);
  const meta = response.usageMetadata;
  const usage = {
    input_tokens: meta.promptTokenCount,
    output_tokens: meta.candidatesTokenCount,
  };
  const finishReason = response.candidates?.[0]?.finishReason;
  return { text: response.text, usage, stop_reason: finishReason };
}

async function callLlm(provider, model, prompt, { effort } = {}) {
  const outputSchema = loadTranslationSchema(provider);
  const start = Date.now();

  const providerCalls = {
    [PROVIDER_ANTHROPIC]: [makeClaudeClient, (client) => callClaude(client, prompt, { model, outputSchema, effort })],
    [PROVIDER_GOOGLE]: [makeGeminiClient, (client) => callGemini(client, prompt, { model, outputSchema })],
  };
  if (!providerCalls[provider]) {
    throw new Error(`Unknown provider: ${provider}`);
  }
  const [makeClient, call] = providerCalls[provider];

  const backends = backendOrder();
  let reason = null;
  for (let i = 0; i < backends.length; i++) {
    const backend = backends[i];
    let result;
    try {
      result = await call(makeClient(backend));
    } catch (err) {
      reason = fallbackReason(err);
      const isLast = i === backends.length - 1;
      if (!reason || isLast) {
        throw err;
      }
      console.warn(`${provider} via ${backend} failed (${reason}); falling back to ${backends[i + 1]}`);
      continue;
    }
    result.usage.duration_ms = Date.now() - start;
    result.usage.llm_backend = backend;
    if (reason) {
      result.usage.llm_fallback_reason = reason;
    }
    return result;
  }
}


module.exports = {
  callClaude,
  callGemini,
  callLlm,
  fallbackReason,
  makeClaudeClient,
  makeGeminiClient,
  BACKEND_VERTEX,
  BACKEND_DIRECT,
  loadTranslationSchema,
  PROVIDER_ANTHROPIC,
  PROVIDER_GOOGLE,
  DEFAULT_MAX_TOKENS,
};

File: cloud-run/extract/llm.py
import base64
import json
import logging
import os
import time
from http import HTTPStatus
from pathlib import Path

from dotenv import load_dotenv

load_dotenv(Path(__file__).resolve().parent.parent.parent / ".env")

import anthropic
import google.auth.exceptions
from google import genai
from google.genai import errors as genai_errors

logging.getLogger("google_genai.models").setLevel(logging.ERROR)
logger = logging.getLogger(__name__)

PROVIDER_ANTHROPIC = "anthropic"
PROVIDER_GOOGLE = "google"
LLM_TIMEOUT_SECONDS = 240

# Backends: Vertex AI (service account auth, no API keys) or the direct
# provider APIs (API keys from Secret Manager). LLM_BACKEND picks the order;
# vertex-first falls back to direct while the Vertex migration is verified.
BACKEND_VERTEX = "vertex"
BACKEND_DIRECT = "direct"
LLM_BACKEND_ENV_VAR = "LLM_BACKEND"
BACKEND_MODE_VERTEX_FIRST = "vertex-first"
BACKEND_MODE_VERTEX_ONLY = "vertex-only"
BACKEND_MODE_DIRECT_ONLY = "direct-only"
BACKEND_ORDER = {
    BACKEND_MODE_VERTEX_FIRST: [BACKEND_VERTEX, BACKEND_DIRECT],
    BACKEND_MODE_VERTEX_ONLY: [BACKEND_VERTEX],
    BACKEND_MODE_DIRECT_ONLY: [BACKEND_DIRECT],
}
VERTEX_PROJECT_ENV_VAR = "VERTEX_PROJECT_ID"
VERTEX_LOCATION_ENV_VAR = "VERTEX_LOCATION"
DEFAULT_VERTEX_LOCATION = "us"

# Fall back only on errors raised before any generation work (auth, model not
# enabled/available, quota). Timeouts, 5xx, and bad output are not retried on
# the other backend: they can come after minutes of work and would double
# cost and latency while hiding the real failure.
FALLBACK_STATUS_CODES = {
    HTTPStatus.UNAUTHORIZED,
    HTTPStatus.FORBIDDEN,
    HTTPStatus.NOT_FOUND,
    HTTPStatus.TOO_MANY_REQUESTS,
}


class VertexNotConfiguredError(RuntimeError):
    """VERTEX_PROJECT_ID is unset, so there's no Vertex project to call."""


def _vertex_settings():
    project = os.environ.get(VERTEX_PROJECT_ENV_VAR)
    if not project:
        raise VertexNotConfiguredError(f"{VERTEX_PROJECT_ENV_VAR} is not set")
    return project, os.environ.get(VERTEX_LOCATION_ENV_VAR, DEFAULT_VERTEX_LOCATION)


def make_claude_client(backend):
    if backend == BACKEND_VERTEX:
        project, location = _vertex_settings()
        return anthropic.AnthropicVertex(project_id=project, region=location, timeout=LLM_TIMEOUT_SECONDS)
    return anthropic.Anthropic(timeout=LLM_TIMEOUT_SECONDS)


def make_gemini_client(backend):
    if backend == BACKEND_VERTEX:
        project, location = _vertex_settings()
        return genai.Client(vertexai=True, project=project, location=location)
    return genai.Client(api_key=os.environ.get("GEMINI_API_KEY"))


def fallback_reason(err):
    """Why this error justifies retrying on the direct API, or None if it doesn't."""
    if isinstance(err, VertexNotConfiguredError):
        return "vertex not configured"
    if isinstance(err, google.auth.exceptions.GoogleAuthError):
        return f"credentials: {type(err).__name__}"
    if isinstance(err, anthropic.APIStatusError) and err.status_code in FALLBACK_STATUS_CODES:
        return f"HTTP {err.status_code}: {type(err).__name__}"
    if isinstance(err, genai_errors.APIError) and err.code in FALLBACK_STATUS_CODES:
        return f"HTTP {err.code}: {err.status}"
    return None


def backend_order():
    mode = os.environ.get(LLM_BACKEND_ENV_VAR, BACKEND_MODE_VERTEX_FIRST)
    if mode not in BACKEND_ORDER:
        raise ValueError(f"Unknown {LLM_BACKEND_ENV_VAR}: {mode}")
    return BACKEND_ORDER[mode]

SCHEMA_DIR = Path(__file__).resolve().parent
SCHEMA_PATHS = {
    PROVIDER_ANTHROPIC: SCHEMA_DIR / "extraction-schema-claude.json",
    PROVIDER_GOOGLE: SCHEMA_DIR / "extraction-schema-gemini.json",
}


def call_claude(client, prompt, *, model="claude-sonnet-4-6", max_tokens=65536, system=None, pdf_base64=None, output_schema=None):
    content = []

    if pdf_base64:
        content.append({
            "type": "document",
            "source": {"type": "base64", "media_type": "application/pdf", "data": pdf_base64},
        })

    content.append({"type": "text", "text": prompt})

    kwargs = {"model": model, "max_tokens": max_tokens, "messages": [{"role": "user", "content": content}]}
    if system:
        kwargs["system"] = system
    if output_schema:
        kwargs["output_config"] = {
            "format": {
                "type": "json_schema",
                "schema": output_schema,
            }
        }

    response = client.messages.create(**kwargs)
    usage = {
        "input_tokens": response.usage.input_tokens,
        "output_tokens": response.usage.output_tokens,
    }
    text_block = next(b for b in response.content if b.type == "text")
    return text_block.text, usage


def call_gemini(client, prompt, *, model="gemini-3.5-flash", pdf_base64=None, output_schema=None):
    parts = []

    if pdf_base64:
        data = base64.b64decode(pdf_base64) if isinstance(pdf_base64, str) else pdf_base64
        parts.append(genai.types.Part.from_bytes(data=data, mime_type="application/pdf"))

    parts.append(genai.types.Part.from_text(text=prompt))

    kwargs = {"model": model, "contents": [genai.types.Content(role="user", parts=parts)]}
    if output_schema:
        kwargs["config"] = genai.types.GenerateContentConfig(
            response_mime_type="application/json",
            response_schema=output_schema,
        )

    response = client.models.generate_content(**kwargs)
    usage_meta = response.usage_metadata
    usage = {
        "input_tokens": usage_meta.prompt_token_count,
        "output_tokens": usage_meta.candidates_token_count,
    }
    return response.text, usage


def load_extraction_schema(provider):
    path = SCHEMA_PATHS.get(provider)
    if not path:
        raise ValueError(f"No extraction schema for provider: {provider}")
    return json.loads(path.read_text())


def call_llm(provider, model, prompt, pdf_base64):
    """Call the provider, trying backends in LLM_BACKEND order.

    Returns (text, usage). usage includes duration_ms (total, across any
    fallback), llm_backend ("vertex" or "direct"), and llm_fallback_reason
    when Vertex failed and the direct API answered.
    """
    output_schema = load_extraction_schema(provider)
    # Built per call (not at import) so tests can patch the functions.
    provider_calls = {
        PROVIDER_ANTHROPIC: (make_claude_client, call_claude),
        PROVIDER_GOOGLE: (make_gemini_client, call_gemini),
    }
    if provider not in provider_calls:
        raise ValueError(f"Unknown provider: {provider}")
    make_client, call = provider_calls[provider]

    start = time.monotonic()
    backends = backend_order()
    reason = None
    for i, backend in enumerate(backends):
        try:
            text, usage = call(make_client(backend), prompt, model=model, pdf_base64=pdf_base64, output_schema=output_schema)
        except Exception as err:
            reason = fallback_reason(err)
            is_last = i == len(backends) - 1
            if reason is None or is_last:
                raise
            logger.warning("%s via %s failed (%s); falling back to %s",
                           provider, backend, reason, backends[i + 1])
            continue
        usage["duration_ms"] = round((time.monotonic() - start) * 1000)
        usage["llm_backend"] = backend
        if reason:
            usage["llm_fallback_reason"] = reason
        return text, usage
